### PR TITLE
bpo-43392: Optimize repeated calls to `__import__()`

### PR DESCRIPTION
**Pull request by Kronuz**
Implements a two steps check in `importlib._bootstrap._find_and_load()` to avoid locking when the module has been already imported and it's ready.

---

Using `importlib.__import__()`, after this, does show a big difference:

Before:
```
$ ./python -c 'import timeit; print(timeit.timeit("__import__(\"timeit\")", setup="from importlib import __import__"))'
15.92248619502061
```

After:
```
$ ./python -c 'import timeit; print(timeit.timeit("__import__(\"timeit\")", setup="from importlib import __import__"))'
1.206068897008663
```

---

https://bugs.python.org/issue43392

Automerge-Triggered-By: GH:pitrou

### CODE REVIEW

File: Lib/importlib/_bootstrap.py
Comment (pitrou) [inline Lib/importlib/_bootstrap.py:1050]: This optimization looks a lot like what is done in `import_ensure_initialized` in `Python/import.c`, can I ask you to add similar comments to those in that function?

File: Lib/importlib/_bootstrap.py
Comment (pitrou) [inline Lib/importlib/_bootstrap.py]: Is there a reason for taking so much vertical space?
```python
if (module is _NEEDS_LOADING or
    getattr(getattr(module, "__spec__", None), "_initializing", False)):
    # ...
```

**Review by pitrou — CHANGES_REQUESTED** [review]
The optimization looks reasonable in the principle, as there's already a similar one on the C side.

File: Lib/importlib/_bootstrap.py
Comment (Kronuz) [reply Lib/importlib/_bootstrap.py:1050]: Done! I added a few comments about the optimizations and tweaked the code as suggested.

**Comment by brettcannon** [comment]
@Kronuz if this is ready for review again then please see the instructions at https://github.com/python/cpython/pull/24735#issuecomment-821296686 on how to request a re-review.

**Comment by Kronuz** [comment]
Oh! @brettcannon, yes I have made the requested changes; please review again. I just missed that part of the instructions.

**Comment by Kronuz** [comment]
Didn't close it... reopening 🤪

**Comment by Kronuz** [comment]
@pitrou, I've addressed the suggestions and made the changes, however I'm not sure if I'm doing something wrong or why this doesn't show as all green and instead has 2 workflows awaiting approval and the "_Changes requested_" status. Please let me know if I'm missing something.

**Comment by pitrou** [comment]
@Kronuz You just had to post the phrase "I have made the requested changes; please review again" as mentioned by the bot above :)

As for the CI steps, they must be approved by a core developer since this is your first PR.

**Comment by pitrou** [comment]
@pablogsal @iritkatriel Do you want to take a look at this?

File: Misc/NEWS.d/next/Library/2021-03-03-13-32-37.bpo-43392.QQumou.rst
Comment (brettcannon) [inline Misc/NEWS.d/next/Library/2021-03-03-13-32-37.bpo-43392.QQumou.rst]:
```suggestion
:func:`importlib._bootstrap._find_and_load` now implements a two-step
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # putting the new module in sys.modules.
```

**Review by brettcannon — CHANGES_REQUESTED** [review]
Minor comment tweaks, but the PR doesn't allow for modification by me so I can't fix them.

Otherwise `importlib.h` is out-of-date. Let's give people a week to add any review they may want, otherwise I will merge this once `importlib.h` is no longer a merge conflict.

**Comment by Kronuz** [comment]
I have made the requested changes; please review again

**Review by pitrou — APPROVED** [review]
This looks good to me.

**Pull request merged**
